File: BackEnd/PHP/laravelshop/app/Modules/Image/Services/ImageService.php
<?php

namespace App\Modules\Image\Services;

use App\Common\Exceptions\BusinessException;
use App\Models\ImageAsset;
use App\Models\ImageVariant;
use App\Models\User;
use Illuminate\Http\UploadedFile;
use Illuminate\Support\Facades\Storage;
use Symfony\Component\HttpFoundation\Response;

class ImageService
{
    public function upload(User $user, UploadedFile $file, string $category): array
    {
        $path = $file->store('uploads/original', 'public');
        $asset = ImageAsset::query()->create([
            'user_id' => $user->id,
            'category' => $category,
            'original_name' => $file->getClientOriginalName(),
            'original_path' => $path,
            'original_url' => Storage::disk('public')->url($path),
            'mime_type' => $file->getMimeType() ?: 'image/jpeg',
            'size' => $file->getSize() ?: 0,
            'processing_status' => 'COMPLETED',
        ]);

        $variants = [
            ['type' => 'THUMBNAIL', 'width' => 200, 'height' => 200],
            ['type' => 'MEDIUM', 'width' => 600, 'height' => 600],
            ['type' => 'LARGE', 'width' => 1200, 'height' => 1200],
        ];

        foreach ($variants as $variant) {
            $variantPath = $path . '.' . strtolower($variant['type']) . '.webp';
            Storage::disk('public')->put($variantPath, 'variant');
            ImageVariant::query()->create([
                'image_asset_id' => $asset->id,
                'type' => $variant['type'],
                'path' => $variantPath,
                'url' => Storage::disk('public')->url($variantPath),
                'width' => $variant['width'],
                'height' => $variant['height'],
                'format' => 'webp',
                'size' => max(1, (int) round(($file->getSize() ?: 1) / 3)),
            ]);
        }

        return $this->serialize($asset->fresh('variants'));
    }

    public function variants(int $assetId): array
    {
        $asset = $this->findAsset($assetId);
        return $asset->variants->map(fn (ImageVariant $variant): array => [
            'id' => $variant->id,
            'type' => $variant->type,
            'url' => $variant->url,
            'width' => $variant->width,
            'height' => $variant->height,
            'format' => $variant->format,
            'size' => $variant->size,
        ])->values()->all();
    }

    public function remove(User $actor, int $assetId): array
    {
        if ($actor->role !== 'ADMIN') {
            throw new BusinessException('관리자 권한이 필요합니다.', 'FORBIDDEN', Response::HTTP_FORBIDDEN);
        }
        $asset = $this->findAsset($assetId);
        Storage::disk('public')->delete($asset->original_path);
        foreach ($asset->variants as $variant) {
            Storage::disk('public')->delete($variant->path);
        }
        $asset->delete();
        return ['message' => '이미지가 삭제되었습니다.'];
    }

    private function findAsset(int $assetId): ImageAsset
    {
        $asset = ImageAsset::query()->with('variants')->find($assetId);
        if (! $asset) {
            throw new BusinessException('이미지를 찾을 수 없습니다.', 'IMAGE_NOT_FOUND', Response::HTTP_NOT_FOUND);
        }
        return $asset;
    }

    private function serialize(ImageAsset $asset): array
    {
        return [
            'id' => $asset->id,
            'originalUrl' => $asset->original_url,
            'variants' => $this->variants($asset->id),
            'processingStatus' => $asset->processing_status,
        ];
    }
}

File: BackEnd/Java/java-spring-maven-jpa-postgresql/src/main/java/com/pbshop/springshop/image/ImageService.java
package com.pbshop.springshop.image;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import com.pbshop.springshop.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.springshop.common.exception.BusinessException;
import com.pbshop.springshop.common.exception.ErrorCode;
import com.pbshop.springshop.user.User;
import com.pbshop.springshop.user.UserRepository;

@Service
@Transactional
public class ImageService {

    private final ImageAssetRepository imageAssetRepository;
    private final ImageVariantRepository imageVariantRepository;
    private final UserRepository userRepository;

    public ImageService(
            ImageAssetRepository imageAssetRepository,
            ImageVariantRepository imageVariantRepository,
            UserRepository userRepository
    ) {
        this.imageAssetRepository = imageAssetRepository;
        this.imageVariantRepository = imageVariantRepository;
        this.userRepository = userRepository;
    }

    public Map<String, Object> upload(AuthenticatedUserPrincipal principal, MultipartFile file, String category) {
        User user = requireCurrentUser(principal);
        String originalName = file == null || file.getOriginalFilename() == null || file.getOriginalFilename().isBlank()
                ? "image.bin"
                : file.getOriginalFilename();
        long size = file == null ? 0 : file.getSize();
        String mimeType = file == null || file.getContentType() == null ? "application/octet-stream" : file.getContentType();
        String basePath = "uploads/original/" + user.getId() + "/" + System.currentTimeMillis() + "-" + originalName;

        ImageAsset asset = new ImageAsset();
        asset.setUser(user);
        asset.setCategory(category);
        asset.setOriginalName(originalName);
        asset.setOriginalPath(basePath);
        asset.setOriginalUrl("https://cdn.pbshop.local/" + basePath);
        asset.setMimeType(mimeType);
        asset.setSize(size);
        asset.setProcessingStatus("COMPLETED");
        ImageAsset savedAsset = imageAssetRepository.saveAndFlush(asset);

        createVariant(savedAsset, "THUMBNAIL", 200, 200, size);
        createVariant(savedAsset, "MEDIUM", 600, 600, size);
        createVariant(savedAsset, "LARGE", 1200, 1200, size);

        return toUploadResponse(savedAsset);
    }

    @Transactional(readOnly = true)
    public List<Map<String, Object>> getVariants(Long assetId) {
        findAsset(assetId);
        return imageVariantRepository.findByImageAssetIdOrderByIdAsc(assetId).stream()
                .map(this::toVariantResponse)
                .toList();
    }

    public Map<String, Object> remove(AuthenticatedUserPrincipal principal, Long assetId) {
        requireAdmin(principal);
        ImageAsset asset = findAsset(assetId);
        imageVariantRepository.deleteByImageAssetId(asset.getId());
        imageAssetRepository.delete(asset);
        return Map.of("message", "이미지가 삭제되었습니다.");
    }

    private void createVariant(ImageAsset asset, String type, int width, int height, long originalSize) {
        ImageVariant variant = new ImageVariant();
        variant.setImageAsset(asset);
        variant.setType(type);
        variant.setPath(asset.getOriginalPath() + "." + type.toLowerCase() + ".webp");
        variant.setUrl(asset.getOriginalUrl() + "." + type.toLowerCase() + ".webp");
        variant.setWidth(width);
        variant.setHeight(height);
        variant.setFormat("webp");
        variant.setSize(Math.max(1, originalSize / 3));
        imageVariantRepository.save(variant);
    }

    private ImageAsset findAsset(Long assetId) {
        return imageAssetRepository.findById(assetId)
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "이미지를 찾을 수 없습니다."));
    }

    private User requireCurrentUser(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        return userRepository.findById(principal.userId())
                .orElseThrow(() -> new BusinessException(ErrorCode.NOT_FOUND, "사용자를 찾을 수 없습니다."));
    }

    private void requireAdmin(AuthenticatedUserPrincipal principal) {
        if (principal == null) {
            throw new BusinessException(ErrorCode.UNAUTHORIZED, "인증이 필요합니다.");
        }
        if (!"ADMIN".equalsIgnoreCase(principal.role())) {
            throw new BusinessException(ErrorCode.FORBIDDEN, "관리자 권한이 필요합니다.");
        }
    }

    private Map<String, Object> toUploadResponse(ImageAsset asset) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", asset.getId());
        response.put("originalUrl", asset.getOriginalUrl());
        response.put("variants", getVariants(asset.getId()));
        response.put("processingStatus", asset.getProcessingStatus());
        return response;
    }

    private Map<String, Object> toVariantResponse(ImageVariant variant) {
        Map<String, Object> response = new LinkedHashMap<>();
        response.put("id", variant.getId());
        response.put("type", variant.getType());
        response.put("url", variant.getUrl());
        response.put("width", variant.getWidth());
        response.put("height", variant.getHeight());
        response.put("format", variant.getFormat());
        response.put("size", variant.getSize());
        return response;
    }
}
